### PR TITLE
gpu(::DataLoader), take II

### PR DESCRIPTION
**Pull request by mcabbott**
Closes https://github.com/FluxML/Flux.jl/pull/2186 , by replacing CuIterator with much the same thing, but owned by Flux.

The advantage of this is that we can control exactly what it does, to match `gpu(x)`. This means converting to Float32, and also working with AMDGPU arrays (and whatever we add next).

Needs tests.

### PR Checklist

- [x] Tests are added
- [x] Entry in NEWS.md
- [x] Documentation, if applicable

### CODE REVIEW

File: docs/src/gpu.md [hunk 1]
```diff
@@ -155,26 +170,14 @@ In order to train the model using the GPU both model and the training data have
...
    gpu_train_loader = Flux.DataLoader((xtrain, ytrain) |> gpu, batchsize = 32)
    ```
-   Note that both `gpu` and `cpu` are smart enough to recurse through tuples and namedtuples. Another possibility is to use [`MLUtils.mapsobs`](https://juliaml.github.io/MLUtils.jl/dev/api/#MLUtils.mapobs) to push the data movement invocation into the background thread:
+   Note that both `gpu` and `cpu` are smart enough to recurse through tuples and namedtuples.
+   Another possibility is to use [`MLUtils.mapsobs`](https://juliaml.github.io/MLUtils.jl/dev/api/#MLUtils.mapobs) to push the data movement invocation into the background thread:
    ```julia
    using MLUtils: mapobs
    # ...
    gpu_train_loader = Flux.DataLoader(mapobs(gpu, (xtrain, ytrain)), batchsize = 16)
    ```
```
Comment (mcabbott) [inline docs/src/gpu.md:179]: I saw this while updating the docs. Does it work? `mapobs` seems to obey strange rules:
```julia
julia> data = (a=[1 2 3; 4 5 6], b=[7,8,9]);

julia> mapobs(pr, data) |> collect
x = (a = [1, 4], b = 7)
x = (a = [2, 5], b = 8)
x = (a = [3, 6], b = 9)
3-element Vector{Any}:
 (a = [1, 4], b = 7)
 (a = [2, 5], b = 8)
 (a = [3, 6], b = 9)

julia> DataLoader(mapobs(pr, data); batchsize=2) |> collect
x = (a = [1 2; 4 5], b = [7, 8])
x = (a = [3; 6;;], b = [9])
2-element Vector{NamedTuple{(:a, :b), Tuple{Matrix{Int64}, Vector{Int64}}}}:
 (a = [1 2; 4 5], b = [7, 8])
 (a = [3; 6;;], b = [9])
```
But if it does work, this might be the obvious thing for `gpu(::DataLoader)` to do.

Comment (mcabbott) [reply docs/src/gpu.md:179]: Why it's written up as a method of moving the entire dataset to the GPU is confusing. Seems to have been there from the start, and nobody noticed?

https://github.com/FluxML/Flux.jl/pull/1980/files#diff-813fc68aeb136c3af32fae13632adf27c24229ce9b56b151d13b836db8a7aa2dR117-R122
